### ISSUE TITLE
App params - options and see request?

### ISSUE THREAD

**Issue by aspot** [merdok/homebridge-webos-tv#335]
Do you know if there documentation listed anywhere on what the **available** params are for each app?

I just see WebOS mentioning it here: https://www.webosose.org/docs/reference/ls2-api/com-webos-service-applicationmanager/#params

The only example I see outside of that is at the top of the page in the description: https://www.webosose.org/docs/reference/ls2-api/com-webos-service-applicationmanager/#overview-of-the-api which shows the ContentTarget.

In WebOS, you can pin a YouTube channel alongside other apps on the Home pinwheel. I'm trying to reverse engineer what the ContentTarget needs to be (or other parameter?) to emulate that with homebridge-webos-tv. Enabling debug mode with verbose/detailed logging does not reveal params used in the request when I click the WebOS YouTube channel shortcut - it seems to just reveal the response back from the TV after an action (e.g. open AppID) and not the actual request payload. Not sure if it's possible to grab the request?

**Comment by merdok** [comment]
I do not think that this is a general public info what params each app has, since it would be no use to normal users. Some of them are known like that `contentTarget` which is used by many apps. 

If you enable the `deepDebugLog` in my plugin, then the requests from the TV should also be logged, this is includes a list of apps their full definition, you might wanna check that out if you find any useful info there. 

I also tried to do that at some point but could not find anything useful, maybe i missed something or something changed... if you find anything then please let me know.

**Comment by aspot** [comment]
Thanks for the response. I can't find anything on the internet either, but if I do I will post back here.

I do have `deepDebugLog` enabled and debug mode (`-D`) for Homebridge, but here's the logs I see:

[12/24/2020, 8:17:49 AM] [webostv] [LG65CX] TV setup finished, ready to control tv
[12/24/2020, 8:18:32 AM] [webostv] [LG65CX] Message from TV 
[12/24/2020, 8:18:32 AM] [webostv] [LG65CX]  {
  type: 'utf8',
  utf8Data: '{"type":"response","id":"1f3069e30033","payload":{"subscribed":true,"appId":"com.webos.app.homeconnect","returnValue":true,"windowId":"","processId":""}}'
}
[12/24/2020, 8:18:32 AM] [webostv] [LG65CX] {
  "subscribed": true,
  "appId": "com.webos.app.homeconnect",
  "returnValue": true,
  "windowId": "",
  "processId": ""
}
[12/24/2020, 8:18:32 AM] [webostv] [LG65CX] App launched, current appId: com.webos.app.homeconnect
[12/24/2020, 8:18:53 AM] [webostv] [LG65CX] Message from TV 
[12/24/2020, 8:18:53 AM] [webostv] [LG65CX]  {
  type: 'utf8',
  utf8Data: '{"type":"response","id":"1f3069e30033","payload":{"subscribed":true,"appId":"youtube.leanback.v4","returnValue":true,"windowId":"","processId":""}}'
}
[12/24/2020, 8:18:53 AM] [webostv] [LG65CX] {
  "subscribed": true,
  "appId": "youtube.leanback.v4",
  "returnValue": true,
  "windowId": "",
  "processId": ""
}
[12/24/2020, 8:18:53 AM] [webostv] [LG65CX] App launched, current appId: youtube.leanback.v4

As you can see, it only looks like the response from the TV was logged. I know `appId: youtube.leanback.v4` was launched with some sort of `params` but I just see `[LG65CX] App launched, current appId: youtube.leanback.v4`.

**Comment by merdok** [comment]
@aspot  you have to check before the setup is finished, usually when turning on your TV. At that time a sync is made and all the apps from the TV are retrieved.

**Comment by aspot** [comment]
@merdok Excellent, that did the trick - I was able to see that `"contentTarget": "c={YOUTUBE_CHANNEL_ID}"` worked!

Now to see how to jump to specific Netflix shows... ;)

**Comment by merdok** [comment]
Did you try that for netflix shows:
https://github.com/merdok/homebridge-webos-tv/blob/b5893f73061884e66141a5f40e72b9679d1b74ab/lib/LgTvController.js#L931-L940

The `contentTarget` needs to have a specific format.

**Comment by aspot** [comment]
I've had some time to test around today.

`{
                            "appId": "netflix",
                            "name": "Show",
							"params": {
								"contentTarget": "m=http://api.netflix.com/catalog/titles/movies/{showID}&source_type=4"
							}
                        },`

This works to launch Netflix with the given `{showId}` and it does not have to be a movie even though the URL format would leave you to believe it has to be a movie title. I'm using it for a TV series.

Launching YouTube with `"contentTarget": "v={videoID}"` will launch the video but it will present a prompt saying "Are you sure you want to play this video?" and you have to select Yes or Cancel. And the `"contentTarget": "c={YOUTUBE_CHANNEL_ID}"` does open the channel but does not autoplay any content.

I'm assuming there must be a parameter or query param to append to `contentTarget` to enable autoplay, but haven't tried anything.

Vielen Dank @merdok !

**Comment by merdok** [comment]
Yeah, they added this confirmation prompt in some of the recent updates of the youtube app. This is really annoying since it does not autoplay videos anymore now because of that prompt, but maybe there is a parameter which does skips that prompt...

**Comment by aspot** [comment]
This is a little unrelated, but I've found the homebridge app shortcut doesn't turn on the TV. So if I wanted to automate a "Hey Siri" for example to launch YouTube on the TV, I believe I would need to create a Shortcut which sequentially calls **two** scenes in Homekit - the first to turn on the TV, and the second to launch the app. Creating a single Scene with both these actions does not work because the actions are not synchronous; i.e. it will try to launch the app before the TV is turned on. Not sure if you had any comments or experience with this, but just wanted to throw this out in the aether.

**Comment by merdok** [comment]
You can tell Siri (or directly use the button) to launch YouTube even when the tv is off. This will work as you expect, the tv will turn on and the YouTube App will open.

Your are right that HomeKit does not support that, but I thought about it and build such functionality in my plugin😉
This is also documented in the README at the appButtons configuration fields.

**Comment by aspot** [comment]
Thanks I actually am using the AppButtons! They weren't triggering the TV to fire and I realized in Homekit I was unable to power on the TV going through the Homebridge accessory or the native LG accessory. I had to use the remote to turn the TV on. 

I think for whatever reason it loses its synchronization sometimes. This has only happened once or twice for me so far but the AppButtons are perfect. If it does happen again I'll see if there's anything on the homebridge logs I can share.
